js/autocomplete.js

```diff
@@ -226,99 +221,88 @@
-      let matchingData = [];
-      // Gather all matching data
-      for (let key in data) {
-        if (data.hasOwnProperty(key) && key.toLowerCase().indexOf(val) !== -1) {
-          let entry = {
-            data: data[key],
-            key: key
-          };
-          matchingData.push(entry);
 
-          this.count++;
-        }
-      }
+      let matchingData = data.filter(
+        (entry) => (entry.text || entry.id).toLowerCase().indexOf(selectedValue) !== -1
+      );
+      this.count = matchingData.length;
+
       // Sort
       if (this.options.sortFunction) {
         let sortFunctionBound = (a, b) => {
           return this.options.sortFunction(
-            a.key.toLowerCase(),
-            b.key.toLowerCase(),
-            val.toLowerCase()
+            (a.text || a.id).toLowerCase(),
+            (b.text || b.id).toLowerCase(),
+            selectedValue.toLowerCase()
           );
         };
         matchingData.sort(sortFunctionBound);
       }
       // Limit
       matchingData = matchingData.slice(0, this.options.limit);
+
       // Render
       for (let i = 0; i < matchingData.length; i++) {
         const entry = matchingData[i];
         const item = document.createElement('li');
-        if (!!entry.data) {
+        item.setAttribute('data-id', entry.id);
+        if (entry.image) {
           const img = document.createElement('img');
           img.classList.add('right', 'circle');
-          img.src = entry.data;
+          img.src = entry.image;
           item.appendChild(img);
         }
-        const parts = this._highlightPartialText(val, entry.key);
-        const s = document.createElement('span');
+        const parts = this._highlightPartialText(selectedValue, entry.text || entry.id);
+        const span = document.createElement('span');
         if (this.options.allowUnsafeHTML) {
-          s.innerHTML = parts[0] + '<span class="highlight">' + parts[1] + '</span>' + parts[2];
+          span.innerHTML = parts[0] + '<span class="highlight">' + parts[1] + '</span>' + parts[2];
         } else {
-          s.appendChild(document.createTextNode(parts[0]));
-          if (!!parts[1]) {
+          span.appendChild(document.createTextNode(parts[0]));
+          if (parts[1]) {
             const highlight = document.createElement('span');
             highlight.textContent = parts[1];
             highlight.classList.add('highlight');
-            s.appendChild(highlight);
-            s.appendChild(document.createTextNode(parts[2]));
+            span.appendChild(highlight);
+            span.appendChild(document.createTextNode(parts[2]));
           }
         }
-        item.appendChild(s);
+        item.appendChild(span);
         $(this.container).append(item);
       }
     }
 
-    selectOption(el) {
-      let text = el.text().trim();
-      this.el.value = text;
+    selectOption(id) {
+      const entries = this.options.data.filter((entry) => entry.id == id);
```

Comment:
just a note, IMO this could cost a lot if we have thousands of data there.

Reply:
Yes but filter works in parallel so i dont think its that much of a problem. If you have a better solution to find an entry by the **id** in an array of objects i can implement it.

Reply:
even though it's parallel, it's not multi threaded. AFAIK, JavaScript is single threaded.

---

hmm how about bring back the `data = {}`? If previously like this

```
{
  "Apple": null,
  "Microsoft": null,
  "Google": 'https://placehold.it/250x250'
}
```
we can change it to something like this
```
{
  "10": {"text": "Apple"},
  "id-whatever": {"text": "Microsoft"},
  "wawa": {"text": "Google", "image": "https://placehold.it/250x250"}
}
```
wdyt?

Reply:
if i am allowed to throw in another 2 cts - may I suggest to use a [`Map`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Map#objects_vs._maps)? It looks like a better fitting and more performant data structure for such a scenario 🙂

Reply:
Yes thats both pretty nice ideas. In the case of thousands i would prefer probably Map then. Here are some performance tests i found.
https://gist.github.com/jung-kim/83676b2310c7c2a9c3d8

BUT: I dont think it is a good idea to store/load that many (thousands) of Entries in the client in the first place, so we could implement a dynamic read function which can be customized via a callback and you can use whatever datasource you want (for example ajax or hash-map etc.) and return then an array with the relevant 10 Entries or so. What do you think? Becasue i also need to read from a Database with many entries and want to use the existing database functions for searching and not load everything in the client an search there.

Reply:
huh TIL JavaScript have `Map` object

Dynamic read function.... uh that... that would be autocomplete v2 isn't it? 😅

Reply:
Yes, callback with async read where you can put a database request or whatever and then return the entries, but i have to test it first locally if that would work out 😅